contracts/farm/Farm.sol

```diff
@@ -0,0 +1,197 @@
+    }
+
+    function deposit(address _lpToken, uint256 _amount) external override whenNotPaused {
+        require(_lpToken != address(0), "Zero address");
+        require(_amount > 0, "Zero amount");
+
+        updatePoolReward(_lpToken);
+
+        PoolInfo storage pool = poolInfos[_lpToken];
+        StakeInfo storage stake = stakes[_lpToken][msg.sender];
+
+        if (stake.amount > 0) {
+            uint256 pending = (stake.amount * pool.accRewardPerShare) /
+                1e12 -
+                stake.rewardDebt;
+            if (pending > 0) {
+                address(rewardToken).safeTransfer(msg.sender, pending);
+                emit LogRewardPaid(msg.sender, _lpToken, pending);
+            }
+        }
+
+        stake.amount += _amount;
+        stake.rewardDebt = (stake.amount * pool.accRewardPerShare) / 1e12;
+
+        _lpToken.safeTransferFrom(msg.sender, address(this), _amount);
+        lpTokenBalances[_lpToken] += _amount;
+
+        emit LogDeposit(msg.sender, _lpToken, _amount);
+    }
+
+    function withdraw(
+        address _lpToken,
+        uint256 _amount
+    ) external override whenNotPaused {
+        require(_lpToken != address(0), "Zero address");
+        PoolInfo storage pool = poolInfos[_lpToken];
+        StakeInfo storage stake = stakes[_lpToken][msg.sender];
+
+        require(stake.amount >= _amount, "Not sufficient amount");
+
+        updatePoolReward(_lpToken);
+
+        uint256 pending = (stake.amount * pool.accRewardPerShare) /
+            1e12 -
+            stake.rewardDebt;
+        if (pending > 0) {
+            address(rewardToken).safeTransfer(msg.sender, pending);
+            emit LogRewardPaid(msg.sender, _lpToken, pending);
+        }
+
+        if (_amount > 0) {
+            stake.amount -= _amount;
+            _lpToken.safeTransfer(msg.sender, _amount);
+            lpTokenBalances[_lpToken] -= _amount;
+        }
+
+        stake.rewardDebt = (stake.amount * pool.accRewardPerShare) / 1e12;
+
+        emit LogWithdraw(msg.sender, _lpToken, _amount);
+    }
+
+    function emergencyWithdraw(address _lpToken) override external whenPaused {
+        StakeInfo storage stake = stakes[_lpToken][msg.sender];
+        require(stake.amount > 0, "Zero balance");
+
+        uint256 amount = stake.amount;
+
+        stake.amount = 0;
+        stake.rewardDebt = 0;
+
+        _lpToken.safeTransfer(msg.sender, amount);
+        lpTokenBalances[_lpToken] -= amount;
+
+        emit LogWithdraw(msg.sender,  _lpToken, amount);
+    }
+
+    function setRewardPerBlock(uint256 _rewardPerBlock) override external onlyOwner whenNotPaused {
+        rewardPerBlock = _rewardPerBlock;
+    }
+
+    function pause() external override onlyOwner {
+        _pause();
+    }
+
+
+    function unpause() external override onlyOwner {
+        _unpause();
+    }
+
+
+    function updatePoolReward(address _lpToken) public override {
+        PoolInfo storage pool = poolInfos[_lpToken];
```

Comment:
we won't have here revert for zero token for example. Thats why I recommend to move all code of this function to internal function _updatePoolReward. And add verifications to updatePoolReward. It will allow to keep code protected when _updatePoolReward is called externally in updatePoolReward and do not duplicate verifications when _updatePoolReward is called internally for example in deposit function